src/Cooked/MockChain/Balancing.hs

```diff
@@ -1,409 +1,326 @@
-  TxSkel ->
-  Fee ->
-  Set Api.TxOutRef ->
-  Either MockChainError Fee
-estimateTxSkelFee params managedData managedTxOuts managedValidators skel fees collateralIns = do
-  txBodyContent <-
-    left MCEGenerationError $
-      generateBodyContent fees collateralIns params managedData managedTxOuts managedValidators skel
-  let nkeys = Cardano.estimateTransactionKeyWitnessCount txBodyContent
-  txBody <-
-    left (MCEGenerationError . TxBodyError "Error creating body when estimating fees") $
-      Cardano.createAndValidateTransactionBody Cardano.ShelleyBasedEraConway txBodyContent
-  case Cardano.evaluateTransactionFee Cardano.ShelleyBasedEraConway (Emulator.pEmulatorPParams params) txBody nkeys 0 of
-    Emulator.Coin fee -> pure $ Fee fee
+-- | This computes the minimum and maximum possible fee a transaction can cost
+-- based on the current protocol parameters
+getMinAndMaxFee :: (MonadBlockChainBalancing m) => m (Fee, Fee)
+getMinAndMaxFee = do
+  -- Default parameters in case they are not present. It is unclear when/if this
+  -- could actually happen though.
+  let defMaxTxExecutionUnits =
+        Cardano.ExecutionUnits {executionSteps = 10_000_000_000, executionMemory = 14_000_000}
+      defExecutionUnitPrices =
+        Cardano.ExecutionUnitPrices {priceExecutionSteps = 721 Rat.% 10_000_000, priceExecutionMemory = 577 Rat.% 10_000}
+  -- Parameters necessary to compute the maximum possible fee for a transaction
+  params <- Emulator.pProtocolParams <$> getParams
+  let maxTxSize = toInteger $ Cardano.protocolParamMaxTxSize params
+      Emulator.Coin txFeePerByte = Cardano.protocolParamTxFeePerByte params
+      Emulator.Coin txFeeFixed = Cardano.protocolParamTxFeeFixed params
+      Cardano.ExecutionUnitPrices priceESteps priceEMem = fromMaybe defExecutionUnitPrices $ Cardano.protocolParamPrices params
+      Cardano.ExecutionUnits (toInteger -> eSteps) (toInteger -> eMem) = fromMaybe defMaxTxExecutionUnits $ Cardano.protocolParamMaxTxExUnits params
+  -- Final fee accounts for the size of the transaction and the units consumed
+  -- by the execution of scripts from the transaction
+  let sizeFees = txFeeFixed + (maxTxSize * txFeePerByte)
+      eStepsFees = (eSteps * Rat.numerator priceESteps) `div` Rat.denominator priceESteps
+      eMemFees = (eMem * Rat.numerator priceEMem) `div` Rat.denominator priceEMem
+  return (txFeeFixed, sizeFees + eStepsFees + eMemFees)
 
--- TODO: improve our collateral mechanism
+-- | Computes optimal fee for a given skeleton and balances it around those fees.
+-- This uses a dychotomic search for an optimal "balanceable around" fee.
+computeFeeAndBalance :: (MonadBlockChainBalancing m) => Wallet -> Fee -> Fee -> Collaterals -> BalancingOutputs -> Wallet -> TxSkel -> m (TxSkel, Fee, Collaterals)
+computeFeeAndBalance _ minFee maxFee _ _ _ _
+  | minFee > maxFee =
+      throwError $ FailWith "Unreachable case, please report a bug at https://github.com/tweag/cooked-validators/issues"
```

Comment:
Why is this error message generic with a link while most if not all of the others give a reason?

Reply:
Because this should never happen unless the user has called this function by hand with an empty interval